plugins/signals-node/src/SignalService.ts

```diff
@@ -0,0 +1,256 @@
+/*
+ * Copyright 2023 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import {
+  EventBroker,
+  EventParams,
+  EventSubscriber,
+} from '@backstage/plugin-events-node';
+import {
+  ServiceOptions,
+  SignalConnection,
+  SignalEventBrokerPayload,
+} from './types';
+import { RawData, WebSocket, WebSocketServer } from 'ws';
+import { IncomingMessage } from 'http';
+import { v4 as uuid } from 'uuid';
+import { JsonObject } from '@backstage/types';
+import {
+  BackstageIdentityResponse,
+  IdentityApi,
+  IdentityApiGetIdentityRequest,
+} from '@backstage/plugin-auth-node';
+import { Duplex } from 'stream';
+import { LoggerService } from '@backstage/backend-plugin-api';
+
+/** @public */
+export class SignalService implements EventSubscriber {
+  private connections: Map<string, SignalConnection> = new Map<
+    string,
+    SignalConnection
+  >();
+  private eventBroker?: EventBroker;
+  private logger: LoggerService;
+  private identity: IdentityApi;
+  private server: WebSocketServer;
+
+  static create(options: ServiceOptions) {
+    return new SignalService(options);
+  }
+
+  private constructor(options: ServiceOptions) {
+    ({
+      eventBroker: this.eventBroker,
+      logger: this.logger,
+      identity: this.identity,
+    } = options);
+
+    this.server = new WebSocketServer({
+      noServer: true,
+      clientTracking: false,
+    });
+
+    this.eventBroker?.subscribe(this);
+  }
+
+  /**
+   * Handles request upgrade to websocket and adds the connection to internal
+   * list for publish/subscribe functionality
+   * @param req - Request
+   */
+  handleUpgrade = async (
```

Comment:
I don't think the signal service needs to do this much. All of the websocket handling and event receiving we can just move over to the backend instead right? The backend would be responsible for subscribing to the appropriate events from the `EventBroker` and forward them on WebSocket connections, while all the signal service does is to publish messages to the `EventBroker`.

Reply:
Hmm. That kind of makes the signal service just a wrapper around eventbroker? And it would prevent developers to change the upgrade / connection management functionality in their own signal service implementation (for example if they want to use socket.io or something else).

Reply:
Left this as it is but added also the server instance to the parameters. This allows devs to use for example socket.io for the signals just by overriding the default service. I think this would not be possible if we move everything to the `signal-backend`. What do you think?

Reply:
Ah, tbh I don't think we need to provide configurability of the transport mechanism, at least not to swap it out completely. I believe we can be opinionated here and just we provide enough fallbacks to support most deployment scenarios. If individual pieces need to be extended I think we could add that in the future, for example if any extra headers are required for plain HTTP transports.

Reply:
Okay, I will adjust this PR accordingly when I have time